Bugfix/biagas/meshplot and extra lines

biagas opened the pull request:
Resolves #3325

Ensure VTK_LINE cells are processed and put in the lines output for MeshPlot.
Ensure VTK_LINE cells are removed from the opaque polys output for MeshPlot.
This wasn't necessary prior to VTK8 because we were able to pass the lines and polys to the renderer in the same dataset. Now lines are separated from the surface polys in avtMeshFilter.

- [x] Bug fix
- [ ] New feature
- [ ] New Documentation

### How Has This Been Tested?
I ran the full regression suite and rebaselined images that changed as a result.

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added debugging support to my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added any new baselines to the repo


*Don't forget to squash merge when this pull request is approved*

biagas committed:
Fix bug with mesh plot not displaying disconnected line cells.
Ensure VTK_LINE cells are processed and put in the lines output for MeshPlot.
Ensure VTK_LINE cells are removed from the opaque polys output for MeshPlot.

biagas committed:
Update baselines due to changes in Mesh Plot.

markcmiller86 commented on src/visit_vtk/full/vtkLinesFromOriginalCells.C:
```diff
@@ -255,6 +258,22 @@ int vtkLinesFromOriginalCells::RequestData(
               } // if unique edge
           } // for all edge points
       }//for all edges of cell
+
```
So, in the case there  are no split cells so no "avtOriginalCellNumbers", there is an early exit @ line 166 and the logic here will never get executed.

What about the case where we have VTK_LINE primitives in a mesh that did NOT undergoe cell replacement? Maybe that just works fine and this case was the only failing case?

biagas replied:
Good point. I will double-check whether VTK_LINE primitives are still part of the output in this case.

biagas replied:
vtkExtractEdges filter (used in the case where no avtOriginalCellNumbers array is present) does NOT preserve VTK_LINE primitives.  I'm not sure whether its better to sublcass vtkExtractEdges to preserve the lines, or simply resurrect them in our filter after vtkExtractEdges is done.

markcmiller86 replied:
Kinda depends I think on what vtkExtractEdges is intended to do.

From VTK documentation...
>extract cell edges from any type of data
>vtkExtractEdges is a filter to extract edges from a dataset. Edges are extracted as lines or polylines.

It would seem that an edge primitive, `VTK_LINE`, is itself an edge and so *extracting* it from itself is kinda sorta the identity. In other words, I think that is a bug in vtkExtractEdges and that it should return any `VTK_LINE` primitives that it was given.

biagas replied:
Found this email thread:
https://public.kitware.com/pipermail/vtk-developers/2008-May/020685.html
So, VTK_LINE cells do not have edges, so the behavior of vtkExtractEdges is correct for its purposes.

markcmiller86 replied:
I disagree with W's remarks in that email. It may be a correct statement about what he expects the class `vtkExtractEdges` to do. Thats fine, it is probably better named `vtkExtractEdgesFrom2or3DCells` but it I find it awkward to try to support that behavior by backing it up with the mathemtical/topological statement that an edge has no edges or that a face has no faces. If I give you a poly-line (by defn a 1D cell) or a triangle-strip (by defn a 2D cell) and ask you to give me back the *edges* or the *faces*, are you going to return the empty set? I don't think so. Now, make each of those *cells* an entity of size 1 (e.g. a polyline of a single line or a traingle-strip of a single triangle). Do you return empty set then? I don't think so. An edge has *edges* because it *is* an edge. A face has *faces* because it *is* a face. Now, that may not be the desired behavior of `vtkExtractEdges` but to try to justify that behavior with awkard topological statements as opposed to *that is just the way this class is intended to behave* is hard to swallow.

biagas replied:
So, whether or not we agree with it, the intentional behavior in VTK appears to be that given a cell that is comprised of line primitive(s), querying 'number of edges' for that cell will always return 0, and any code that loops on the number of edges of a cell will fail to process lines.  We can implement our own edge extractor that will processes lines as edges.

markcmiller86 commented on src/visit_vtk/full/vtkLinesFromOriginalCells.C:
```diff
@@ -255,6 +258,22 @@ int vtkLinesFromOriginalCells::RequestData(
+
+      // Lines have 0 edges, so won't be processed by above logic.
+      if(cell->GetCellType() == VTK_LINE)
+      {
```
Ignore this @biagas just testing I can indeed add a single comment

biagas committed:
Modified vtkLinesFromOriginalCells to add disconnected line and vertex primitives to output.

Also made the filter work with vtkDataSet input instead of just vtkPolyData.
Added 'UseOriginalCells' flag. When off, it allows this filter to be used
in place of the vtkExtractEdges filter, which loses line and vertex primitives.
Updated avtMeshFilter to use vtkLinesFromOriginalCells in place of vtkExtractEdges.

biagas committed:
Update exodus baseline due to changes in Mesh Plot that now displays
disconnected lines and points.

biagas committed:
Update MeshPlot and its non-glyph mapper to allow setting of pointSize (from pointSizePixels att).

biagas committed:
Update blueprint baselines due to changes in Mesh Plot now rendering points
from its non-glyphing mapper.

biagas commented:
@markcmiller86  I updated vtkLinesFromOriginalCells to add any VTK_LINE, VTK_POLY_LINE, VTK_VERTEX, and VTK_POLY_VERTEX cells.  Removed use vtkExtractEdges altogether since it won't do what we need. What do you think, ready to merge now?

markcmiller86 reviewed (approved):
This is great! Looks like you fixed a number of blank images from blueprint for mesh plots of points. Also, the Exodus baselines look better to me too. MOAB looks about the same. Can't really tell what changed.

BTW...Circle-CI is saying the build failed. It says it built 29 mins ago so am assuming from the most recent changes.

markcmiller86 commented:
If you follow Details link of ci/circleci: build, you can scroll through output to something like

```
[ 13%] Building CXX object visit_vtk/full/CMakeFiles/visit_vtk.dir/vtkMergeTree.C.o
/home/ci/project/src/visit_vtk/full/vtkLinesFromOriginalCells.C: In member function 'virtual int vtkLinesFromOriginalCells::RequestData(vtkInformation*, vtkInformationVector**, vtkInformationVector*)':
/home/ci/project/src/visit_vtk/full/vtkLinesFromOriginalCells.C:157:21: error: 'vtkVisItUtility' has not been declared
   vtkPoints *pts2 = vtkVisItUtility::GetPoints(input);
                     ^
visit_vtk/full/CMakeFiles/visit_vtk.dir/build.make:245: recipe for target 'visit_vtk/full/CMakeFiles/visit_vtk.dir/vtkLinesFromOriginalCells.C.o' failed
make[2]: *** [visit_vtk/full/CMakeFiles/visit_vtk.dir/vtkLinesFromOriginalCells.C.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 13%] Building CXX object viewer/rpc/CMakeFiles/viewerrpc.dir/ViewerState.C.o
```

biagas committed:
add missing include
